### PR TITLE
fix where type promotion

### LINKED ISSUE #73298: torch.where does not handle scalar type correctly
### 🐛 Describe the bug

The following code:
```python
import torch

ab = torch.zeros(5, dtype=torch.bool)
af = torch.zeros(5, dtype=torch.float)
ad = torch.zeros(5, dtype=torch.double)

# This fail as expected
torch.where(ab, af, ad)  # RuntimeError: expected scalar type float but found double

# This works as expected
torch.where(ab, ad, 3.0)

# WTF are you doing!
torch.where(ab, af, 3.0)  # RuntimeError: expected scalar type float but found double
```
I am expecting `torch.where(ab, af, 3.0)` to work fine, because PyTorch only support double scalar, not float scalar. It does not make any sense to require the user to pass a float scalar as argument.

### Versions

```
Collecting environment information...
PyTorch version: 1.12.0a0+gitec8d677
Is debug build: True
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 11.2.0
Clang version: 13.0.1
CMake version: version 3.22.2
Libc version: glibc-2.35

Python version: 3.10.2 (main, Jan 15 2022, 19:56:27) [GCC 11.1.0] (64-bit runtime)
Python platform: Linux-5.16.9-arch1-1-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.55
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 510.54
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.3.1
/usr/lib/libcudnn_adv_infer.so.8.3.1
/usr/lib/libcudnn_adv_train.so.8.3.1
/usr/lib/libcudnn_cnn_infer.so.8.3.1
/usr/lib/libcudnn_cnn_train.so.8.3.1
/usr/lib/libcudnn_ops_infer.so.8.3.1
/usr/lib/libcudnn_ops_train.so.8.3.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.5
[pip3] torch==1.12.0a0+gitec8d677
[pip3] torch-ucc==1.0.0
[pip3] torchani==2.2
[conda] Could not collect
```

cc @nairbv @mruberry

### PR DESCRIPTION
**Pull request by ngimel**
Fixes #73298
I don't know whether `where` kernel actually supports type promotion, nor am I in the mood to find out, so it's manual type promotion. 
Edit: nah, i can't tell TI to "promote to common dtype" because of bool condition, so manual type promotion is our only option. 
I'll see what tests start failing and fix. 
Uses some parts from #62084

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -5167,58 +5121,47 @@ def make_prob_dist(shape, is_contiguous):
...
-                        self.assertEqual(expected, result)
-
-        # Reset the original dtype
-        torch.set_default_dtype(default_dtype)
+    def test_where_scalar_handcrafted_values(self, device):
+        # Tests ScalarxScalar, ScalarxTensor and TensorxScalar
+        # variant of `where` against NumPy version with
+        # handcrafted values.
+        condition_shape = (5, 5)
+        dtypes = (
```
Comment (mruberry) [inline test/test_torch.py:5132]: nit: you can cross-product dtypes using the @dtypes decorator, too, if you want

File: test/test_torch.py [hunk 2]
```diff
@@ -5167,58 +5121,47 @@ def make_prob_dist(shape, is_contiguous):
...
+        shapes = ((), (5,), (1, 5),)
+
+        with torch.no_grad():
+            tensors = (torch.empty(shape, dtype=dtype, device=device).fill_(17)
+                       for shape, dtype in product(shapes, dtypes))
+
+        # Use different values for `x` and `y`
+        # as they are the output values which are compared.
+        x_vals = (True, 3, 7.0, 1 + 0.5j)
+        y_vals = itertools.chain((False, 4, 8.0, 2 + 0.5j), tensors)
```
Comment (mruberry) [inline test/test_torch.py:5146]: Nice asymmetric iteration

File: test/test_torch.py [hunk 3]
```diff
@@ -5167,58 +5121,47 @@ def make_prob_dist(shape, is_contiguous):
...
+            for y in y_vals:
+                condition = torch.empty(*condition_shape, dtype=torch.bool, device=device).bernoulli_()
+                common_dtype = torch.result_type(x, y)
+
+                def check_equal(condition, x, y):
+                    condition_np = condition.cpu().numpy()
+                    x_np = x.cpu().numpy() if isinstance(x, torch.Tensor) else x
+                    y_np = y.cpu().numpy() if isinstance(y, torch.Tensor) else y
+
+                    # NumPy aggressively promotes to double, hence cast to output to correct dtype
```
Comment (mruberry) [inline test/test_torch.py:5157]: Nice comment

**Review by mruberry — APPROVED** [review]
Brilliant!

File: test/test_torch.py (hunk 1 above)
Comment (ngimel) [reply test/test_torch.py:5131]: Even though now instantiating new tests is much cheaper w/o leak checks, logs still look cleaner if there are fewer.

Comment (mruberry) [reply test/test_torch.py:5131]: Yeah that makes sense

**Pull request closed without merging**